subsys: llext: support ARC MPU

laurenmurphyx64 opened the pull request:
Make it so LLEXT allocates a memory region equal or greater to the minimum size supported by the MPU, to ensure no region overlap.

Note: Per discussion in Discord's llext channel, I found out tests aren't currently running on ARC due to the min_ram filter in common. I've filed a feature request [here](https://github.com/zephyrproject-rtos/zephyr/issues/89117) to be able to set different min_ram per architecture. In the meantime, what do you think of the ugly workaround of creating a copy of the each test for ARC with a different min_ram? Example below.

```
...
llext.readonly:
    arch_allow:               # Xtensa needs writable storage
      - arm
      - riscv
    filter: not CONFIG_MPU and not CONFIG_MMU
    extra_conf_files: ['no_mem_protection.conf']
    extra_configs:
      - CONFIG_LLEXT_STORAGE_WRITABLE=n
  llext.readonly.arc:
    arch_allow:
      - arc
    filter: not CONFIG_MPU and not CONFIG_MMU
    extra_conf_files: ['no_mem_protection.conf']
    extra_configs:
      - CONFIG_LLEXT_STORAGE_WRITABLE=n
    min_ram: 128
...
```

laurenmurphyx64 committed:
llext: support ARC MPU

Make it so LLEXT allocates a memory region equal or greater
to the minimum size supported by the MPU, to ensure no region
overlap.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

lyakh commented on subsys/llext/llext_mem.c:
```diff
@@ -77,7 +79,7 @@ static int llext_copy_region(struct llext_loader *ldr, struct llext *ext,
 	 */
 	if (region->sh_flags & SHF_ALLOC) {
-		if (IS_ENABLED(CONFIG_ARM_MPU)) {
+		if (IS_ENABLED(CONFIG_ARM_MPU) | IS_ENABLED(CONFIG_ARC_MPU)) {
```
mathematically it works and will have the same result, but logically you probably wanted `||` here

laurenmurphyx64 replied:
Oops! Fixed

laurenmurphyx64 commented:
EDIT: To make sure we don't break ARC support while submitting PRs for LLEXT going forward, I added a commit duplicating the tests in testcase.yaml for ARC with a test-specific min_ram. Let me know if you want me to remove this and wait for the feature request above to go through.

laurenmurphyx64 commented:
EDIT 2: Okay, I should've known I can't have a test that doesn't run on all integration platforms. So I can't do that solution. Dropped the commit. :disappointed:

pillo79 reviewed (dismissed):
LGTM if you tested it. CI still refuses to do any test on ARC targets, including the newly supported `readonly_mpu`, but this is clearly not a problem of this PR.

pillo79 commented:
> tests aren't currently running on ARC due to the `min_ram` filter in common

While looking at filtering by tag, I saw that the QEMU ARC targets are missing a `ram:` entry in their YAML files and thus are defaulted to a tiny 128k by Twister, but their DTS reserves 4MB of RAM. 
Unless the 128k is really an arch limit, adding `ram: 4096` to the YAML files in `boards/qemu/arc` should allow all of them to run again :slightly_smiling_face:

laurenmurphyx64 committed:
arc: add ram to qemu board yaml so arc tests run

With a missing RAM value in the ARC QEMU board YAML,
Twister assigns it 128. The min_ram filter for the
testcase is higher. The RAM in the ARC QEMU board DTS
is 4 MB. Update ARC QEMU board YAML to include 4 MB
of RAM so tests run on ARC QEMU.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented:
> > tests aren't currently running on ARC due to the `min_ram` filter in common
> 
> While looking at filtering by tag, I saw that the QEMU ARC targets are missing a `ram:` entry in their YAML files and thus are defaulted to a tiny 128k by Twister, but their DTS reserves 4MB of RAM. Unless the 128k is really an arch limit, adding `ram: 4096` to the YAML files in `boards/qemu/arc` should allow all of them to run again 🙂

@pillo79 I'm going to give this a shot! Though I can vouch the LLEXT tests work on ARC QEMU, I have a feeling a lot of other tests aren't currently being run on ARC QEMU and will probably break. But we'll see!

teburd reviewed (approved):
much nicer

ruuddw reviewed (approved):
Thanks, I wasn't aware of (missing) these attributes. Note the 'flash: 4096' is missing as well (but llext or the tests don't care).

laurenmurphyx64 commented:
> Thanks, I wasn't aware of (missing) these attributes. Note the 'flash: 4096' is missing as well (but llext or the tests don't care).

Oh! Sorry, I would've added this to the PR and got it out of the way for you, but the PR was already merged by the time I saw your comment.
